### ISSUE TITLE
Values received via websocket do not match those on the exchange.

### ISSUE THREAD

**Issue by KimWuWei** [ccxt/ccxt#17576]
### Operating System

Debian 11

### Programming Languages

python

### CCXT Version

3.0.67

### Description

Whether I try to obtain the information of one-minute candles at the time of closing using exchange.watch_ohlcv, or if I do it through exchange.watch_trades and then process the result with exchange.build_ohlcvc, I get results that do not match the ones displayed on the exchange's website.

In this case, I am testing it with BTC/USDT on Kucoin, and in all cases, there is a significant difference. Is there any reliable way to ensure the exchange sends the data exactly as they will incorporate it into their historical records?

### Code

```
import asyncio
import ccxt.pro as ccxt
from datetime import datetime


class KucoinOHLCV:
    def __init__(self):
        self.exchange = ccxt.kucoin({
            'enableRateLimit': True,
        })

    async def handle_ohlcv(self, ohlcv):
        print(ohlcv)
        timestamp, open, high, low, close, volume, null = ohlcv[0]
        date = datetime.utcfromtimestamp(timestamp / 1000).strftime('%Y-%m-%d %H:%M:%S')
        print(ohlcv)
        print(f'Date: {date}, Close: {close}')

    async def fetch_ohlcv(self):
        try:
            await self.exchange.load_markets()
            symbol = 'BTC/USDT'
            timeframe = '1m'

            while True:
                trades = await self.exchange.watch_trades(symbol)
                ohlcv = self.exchange.build_ohlcvc(trades, timeframe)
                # ohlcv = await self.exchange.watch_ohlcv(symbol, timeframe)
                if ohlcv:
                    await self.handle_ohlcv(ohlcv)

        finally:
            await self.exchange.close()


if __name__ == '__main__':
    kucoin_ohlcv = KucoinOHLCV()
    asyncio.get_event_loop().run_until_complete(kucoin_ohlcv.fetch_ohlcv())


```

**Comment by KimWuWei** [comment]
Hello,

The issue I see is that there doesn't seem to be a default way to obtain candles from the requested interval through the websocket. Since we have to specify the interval in the watch_ohlcv method, it implies that the obtained information should correspond to the values within that time interval. However, it appears that only instantaneous information is being received.

Is there anyone who uses Kucoin and can comment on how they handle this? Thank you very much.

**Comment by pcriadoperez** [comment]
Hi @KimWuWei , that is correct the watch functions connect through websockets and return only the latest updates.
If you are looking for candles in a requested interval, I'd recommend to use the `fetchOHLCV` (https://docs.ccxt.com/#/?id=ohlcv-candlestick-charts) function where you can specify a timeframe.

**Comment by KimWuWei** [comment]
Thanks @pcriadoperez, I was doing just this before, a fetch call every few seconds, and it worked fine on Binance and Bybit, but in Kucoin I get this error:

![IMG-20230405-WA0014.jpg](https://user-images.githubusercontent.com/71354375/234773424-19aec8b7-29ef-4c17-88e7-122ba1735d83.jpg)

I don't really understand why as it makes just about 7 calls per minute, but after a few hours running the error happens and then it can't even run for a minute using the same api. So I thought the only solution was using websockets.

Isn't using watch_trades + build_ohlcvc supposed to give the same result?

**Comment by pcriadoperez** [comment]
Yes, it sounds like you are hitting kucoins rate limiter. One option is to set `enableRateLimit: true`

And yes watch_trades_ + build_ohlcv should give the same result. Are you seeting newUpdates: false ? 
By default `watch_trades` only returns any new updates, to build the ohlcv you will need an accumulated history of them

**Comment by KimWuWei** [comment]
Thanks, I tried ` enableRateLimit: True ` but get the same error. After this, I also tried to make it so that for the 15-minute candles, the data be only queried if the current minute is a multiple of 15. At first, it seemed to work, but after about two hours, the same error occurred again.

Just tried `setting newUpdates: false` too, but values don't match.

**Comment by ttodua** [comment]
late reply here, but kucoin engine had some updates since then, if anyone still experiences this issue let us know. 

(one minor note: if exchanges provide incorrecly calculated OHLCV data, there is nothing we can do in such case, as CCXT just passes the data from API to user. if there is miscalculation on exchange-side, you can implement your own logic)
